frankenstein/chunk_store.go

```diff
@@ -0,0 +1,379 @@
+		})
+		return err
+	})
+	for _, cc := range resp.ConsumedCapacity {
+		dynamoConsumedCapacity.WithLabelValues("PutItem").
+			Add(float64(*cc.CapacityUnits))
+	}
+	return err
+}
+
+func (c *chunkStore) Get(from, through model.Time, matchers ...*metric.LabelMatcher) ([]Chunk, error) {
+	chunks, err := c.consultChunkIndex(from, through, matchers)
+	if err != nil {
+		return nil, err
+	}
+	return c.fetchChunkData(chunks)
+}
+
+func (c *chunkStore) consultChunkIndex(from, through model.Time, matchers []*metric.LabelMatcher) ([]Chunk, error) {
+	incomingChunkSets := make(chan []Chunk)
+	incomingErrors := make(chan error)
+	hs := hours(from, through)
+	for _, hour := range hs {
+		for _, matcher := range matchers {
+			go func(hour model.Time, matcher *metric.LabelMatcher) {
+				if matcher.Type != metric.Equal {
+					incomingErrors <- fmt.Errorf("%s matcher not supported yet", matcher.Type)
+					return
+				}
+
+				// TODO escaping - this will break if label values contain the separator (:)
+				hashValue := fmt.Sprintf("%s:%s", hour, matcher.Name)
+				rangeMinValue := fmt.Sprintf("%s=%s:%s", matcher.Name, matcher.Value, minChunkID)
+				rangeMaxValue := fmt.Sprintf("%s=%s:%s", matcher.Name, matcher.Value, maxChunkID)
+
+				var resp *dynamodb.QueryOutput
+				err := timeRequest("Query", dynamoRequestDuration, func() error {
+					var err error
+					resp, err = c.dynamodb.Query(&dynamodb.QueryInput{
+						TableName: aws.String(c.tableName),
+						KeyConditions: map[string]*dynamodb.Condition{
+							hashKey: {
+								AttributeValueList: []*dynamodb.AttributeValue{
+									{S: aws.String(hashValue)},
+								},
+								ComparisonOperator: aws.String("EQ"),
+							},
+							rangeKey: {
+								AttributeValueList: []*dynamodb.AttributeValue{
+									{S: aws.String(rangeMinValue)},
+									{S: aws.String(rangeMaxValue)},
+								},
+								ComparisonOperator: aws.String("BETWEEN"),
+							},
+						},
+						ReturnConsumedCapacity: aws.String(dynamodb.ReturnConsumedCapacityTotal),
+					})
+					return err
+				})
+				if resp.ConsumedCapacity != nil {
+					dynamoConsumedCapacity.WithLabelValues("Query").
+						Add(float64(*resp.ConsumedCapacity.CapacityUnits))
+				}
+				if err != nil {
+					log.Errorf("Error querying DynamoDB: %v", err)
+					incomingErrors <- err
+					return
+				}
+
+				chunkSet := []Chunk{}
+				for _, item := range resp.Items {
+					rangeValue := item[rangeKey].S
+					if rangeValue == nil {
+						log.Errorf("Invalid item: %v", item)
+						incomingErrors <- err
+						return
+					}
+					parts := strings.SplitN(*rangeValue, ":", 2)
+					if len(parts) != 2 {
+						log.Errorf("Invalid item: %v", item)
+						incomingErrors <- err
+						return
+					}
+					chunkValue := item[chunkKey].B
+					if rangeValue == nil {
+						log.Errorf("Invalid item: %v", item)
+						incomingErrors <- err
+						return
+					}
+					chunk := Chunk{
+						ID: parts[1],
```

Comment:
We need to populate the `Chunk.Metric` field as well. Otherwise you get back a bunch of chunks without knowing what metric they belong to. Unfortunately there's no good way to get the `Metric` labelset back from DynamoDB yet given the data we write. One idea was to put the labelset into the DynamoDB value field. Another option with less storage overhead (if that matters, given compression and all) would be a separate DynamoDB table for chunk-id -> labelset lookups.

Yet another option could be to store the metric labelset only in the chunk itself (in S3), but that then prevents later optimizations where we need the labelset to decide whether we want to load a chunk from S3 at all.